### PR TITLE
logger: Include thread ID and name in the log messages.

### PR DESCRIPTION
**Pull request by ilya-bobyr**
#### Problem

When debugging multithreaded code, it could be hard to understand which thread generated which messages.
All the threads are actually racing when they are writing into the same log.

#### Summary of Changes

Add thread IDs into all the log messages.

Before:

```
[2023-01-11T03:09:35.349167961Z TRACE solana_streamer::nonblocking::quic] got chunk: Chunk { offset: 1181, bytes: b"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0" }
[2023-01-11T03:09:35.349306664Z TRACE solana_streamer::nonblocking::quic] got chunk: Chunk { offset: 1208, bytes: b"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0" }
[2023-01-11T03:09:35.349336070Z TRACE solana_streamer::nonblocking::quic] chunk is none
[2023-01-11T03:09:35.349346319Z TRACE solana_streamer::nonblocking::quic] sent 1232 byte packet
[2023-01-11T03:09:35.349710900Z DEBUG solana_streamer::nonblocking::quic] stream error: ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" })
[2023-01-11T03:09:35.442150973Z TRACE mio::poll] deregistering event source from poller
[2023-01-11T03:09:36.342416074Z INFO  solana_streamer::nonblocking::quic] Timed out waiting for connection
[2023-01-11T03:09:36.342529619Z TRACE mio::poll] deregistering event source from poller
```

After:

```
[2023-01-11T22:47:14.486961072Z 3 quic-server TRACE solana_streamer::nonblocking::quic] got chunk: Chunk { offset: 1210, bytes: b"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0" }
[2023-01-11T22:47:14.487086700Z 5 quic-server DEBUG rustls::client::tls13] Ticket saved
[2023-01-11T22:47:14.487106988Z 3 quic-server TRACE solana_streamer::nonblocking::quic] chunk is none
[2023-01-11T22:47:14.487122738Z 3 quic-server TRACE solana_streamer::nonblocking::quic] sent 1232 byte packet
[2023-01-11T22:47:14.488232225Z 3 quic-server DEBUG solana_streamer::nonblocking::quic] stream error: ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" })
[2023-01-11T22:47:14.593718961Z 5 quic-server TRACE mio::poll] deregistering event source from poller
[2023-01-11T22:47:15.477674044Z 3 quic-server INFO  solana_streamer::nonblocking::quic] Timed out waiting for connection
[2023-01-11T22:47:15.477785034Z 3 quic-server TRACE mio::poll] deregistering event source from poller
```

### CODE REVIEW

**Comment by brennanwatt** [comment]
I'm personally a fan of this, but I'm wondering if it might be "log bloat" for others... maybe a config option to opt in to logging thread ID?

**Comment by mvines** [comment]
I like the idea. `ThreadId(X) ` is a lot of bloat in each log. I'd rather just see something like `(X) `.

**Comment by mvines** [comment]
> Another interesting proposition would be thread name; someone (I think Michael actually) 

Oh I like this idea so much more! And yeah, it was my OCD problem that caused that

**Comment by brennanwatt** [comment]
> Another interesting proposition would be thread name; someone (I think Michael actually) went through and cleaned up all of our thread names somewhat recently to fit within the 15(?) character limit recently (all of our thread pools have an index at the end to uniquify too). This is useful information, but thread name would add even more characters then `ThreadId(X)` would.

I actually don't mind the log bloat if I get thread name in return :D

**Comment by ilya-bobyr** [comment]
The problem with the option is that when you are looking at a log, it is something that has already happened.
It is not always possible to rerun with the thread IDs now enabled.

As for local runs, indeed, it might be some extra output.
It would be nice if it would be possible to disable it, but, at the same time, is it really that bad?

I've removed the `ThreadId()` wrapping, leaving only the thread ID number.
And added the thread name.

> [...]
> 
> Given that, Brennan's idea about a config option seems interesting. Maybe something we could change on the fly, similar to how we can change rust log level with `solana-validator set-log-filter`

I can look into it, but maybe it is could be a separate PR?
Or would you not accept this without an option of suppressing thread names?
I wonder if anyone would really be disabling them.

**Comment by mvines** [comment]
Another consideration is that logging is already quite expensive. It'd be interesting to bench this change to see what overhead is being added to every log message

**Comment by ilya-bobyr** [comment]
> Another consideration is that logging is already quite expensive. It'd be interesting to bench this change to see what overhead is being added to every log message

Thread ID adds a single `Arc` read.
Name adds another one.
Formatting, should be a bit cheaper, as `env_logger` has a bunch of configuration options.
In the formatter I wrote, they are all hardcoded.

I could look into adding `criterion` into the `logger` package, if this is what you want.
Or would you suggest a different way to measure the overhead?

**Comment by bw-solana** [comment]
> I could look into adding `criterion` into the `logger` package, if this is what you want. Or would you suggest a different way to measure the overhead?

This seems fine to me. `criterion` + performance benches should give us confidence we aren't accidentally introducing significantly more logging latency

**Review by bw-solana — APPROVED** [review]
LGTM pending performance/overhead data

**Comment by t-nelson** [comment]
can this fallback to the thread id only when there's no name set?  also would be great to use the platform tid here instead of the rust one, which cannot be taken to external tools with any reliability

**Comment by ilya-bobyr** [comment]
> can this fallback to the thread id only when there's no name set?

There is a chance that two threads would have the same name, no?
Ids are not that long, so I'm not sure, it would be a big saving.
Plus, it could be easier to process log messages semi-automatically – like grepping, if the IDs are always present.

> also would be great to use the platform tid here instead of the rust one, which cannot be taken to external tools with any reliability

I can look into it.
But as you can see, this PR is already blocked by the requirement to do performance testing.
Adding more features are not going to help get it in :)
And while it is not merged, nobody is benefiting.

In my case, this PR could have saved me time.
I personally do not care about the system thread ID.
So, if someone does have a case where it matters, wouldn't it be more reasonable if they do it?

**Comment by t-nelson** [comment]
> There is a chance that two threads would have the same name, no?

that would be a bug imo

> Adding more features are not going to help get it in :)

it's not a "feature", the rust tid is near useless. shouldn't be particularly expensive to grab the platform tid, they're typically stored in tls anyway

**Comment by ilya-bobyr** [comment]
> > Adding more features are not going to help get it in :)
> 
> it's not a "feature", the rust tid is near useless. shouldn't be particularly expensive to grab the platform tid, they're typically stored in tls anyway

I disagree that they are useless.  They are very useful in my case, as they allowed me to debug an issue, as I could group messages based on the thread that generated them.
At the same time, I have no use for the platform thread ID.

**Comment by t-nelson** [comment]
imagine you see a hot thread in `top` and want to investigate what it's up to

[`gettid`](https://docs.rs/gettid/latest/gettid/) (rip posix name collision) looks like it'll do the trick portably

**Comment by ilya-bobyr** [comment]
> imagine you see a hot thread in `top` and want to investigate what it's up to
> 
> [`gettid`](https://docs.rs/gettid/latest/gettid/) (rip posix name collision) looks like it'll do the trick portably

OK. Thank you for looking for the library as well!
I'll make the change.

**Comment by ilya-bobyr** [comment]
> I'm super happy to have this feature for tests, as I had to write my own implementation #30266 to debug local-cluster tests.
> 
> That said, I totally agree about the concerns about changing every log line in the world. We have performance concerns, we might also break some people's dumb log filter. Can we disable this by default on some env filter and only enable it in tests for now?

While I agree that backward compatibility is important, it also adds complexity and slows things down.  So, it does have downsides.
If there is no existing guarantee that we preserve the current log format, it might be OK to change it across major versions.

I also do not think that additional options are always good (https://github.com/solana-labs/solana/pull/29643#issuecomment-1379583831).
They, again, add complexity.  And here, I think, based on the use cases I've seen, they are unnecessary.  And could even be counterproductive, if you have a log without the IDs from a failed run.

In my mind, the above out wights the inconvenience of someone updating their log filter.

At the same time, I guess, I might need to make sure that the thread name does not contain a space.  As space is currently the separator for the log header elements.

**Comment by t-nelson** [comment]
good news!  no references to "log" in https://docs.solana.com/developing/backwards-compatibility :wink:

**Comment by ilya-bobyr** [comment]
Sorry, I dropped the ball on this somewhat.
Trying to prepare a larger change I'm working on.
I'll get back to this PR in a bit.

**Pull request closed without merging**
